frontend/lib/framework/error-classes-list.cpp

```diff
@@ -537,4 +537,138 @@ void ErrorProcDefExplicitAnonFormal::write(ErrorWriterBase& wr) const {
   wr.code(fn, {formal});
 }
 
+/* parser errors */
+
+void ErrorParsing::write(ErrorWriterBase& wr) const {
+  auto loc = std::get<const Location>(info);
+  auto errorMessage = std::get<std::string>(info);
+  assert(errorMessage.back() != '.' &&
+         "expected no period at end of ErrorParsing message");
+  errorMessage.push_back('.');
+  wr.heading(kind_, type_, loc, errorMessage);
+}
+
+void ErrorCannotAttachPragmas::write(ErrorWriterBase& wr) const {
+  auto loc = std::get<const Location>(info);
+  wr.heading(kind_, type_, loc, "cannot attach pragmas to this statement.");
+}
+
+void ErrorInvalidIndexExpr::write(ErrorWriterBase& wr) const {
+  auto loc = std::get<const Location>(info);
+  wr.heading(kind_, type_, loc, "invalid index expression.");
+}
+
+void ErrorRecordInheritanceNotSupported::write(ErrorWriterBase& wr) const {
+  auto loc = std::get<const Location>(info);
+  wr.heading(kind_, type_, loc,
+             "inheritance is not currently supported for records.");
+  wr.message(
+      "thoughts on what record inheritance should entail can be added to "
+      "https://github.com/chapel-lang/chapel/issues/6851");
+}
+
+void ErrorInvalidNumericLiteral::write(ErrorWriterBase& wr) const {
+  auto loc = std::get<const Location>(info);
+  auto literalStr = std::get<1>(info);
+  auto errMessage = std::get<2>(info);
+  wr.heading(kind_, type_, loc, "error building numeric literal: ", errMessage,
+             ".");
+}
+
+void ErrorMultipleExternalRenaming::write(ErrorWriterBase& wr) const {
+  auto loc = std::get<const Location>(info);
+  wr.heading(
+      kind_, type_, loc,
+      "external symbol renaming can only be applied to one symbol at a time.");
+}
+
+void ErrorPreIncDecOp::write(ErrorWriterBase& wr) const {
+  auto loc = std::get<const Location>(info);
+  auto opSymbol = std::get<1>(info);
+  auto opName = std::get<2>(info);
+  wr.heading(kind_, type_, loc, "'", opSymbol, "' is not a pre-", opName, ".");
+}
+
+void ErrorNewWithoutArgs::write(ErrorWriterBase& wr) const {
+  auto loc = std::get<const Location>(info);
+  wr.heading(kind_, type_, loc,
+             "type in 'new' expression is missing its argument list.");
+}
+
+void ErrorUseImportNeedsModule::write(ErrorWriterBase& wr) const {
+  auto loc = std::get<const Location>(info);
+  auto isImport = std::get<bool>(info);
+  std::string useOrImport = isImport ? "import" : "use";
+  wr.heading(kind_, type_, loc, "'", useOrImport,
+             "' statements must refer to module", (isImport ? "" : " or enum"),
+             " symbols.");
+  wr.message("(e.g., '", useOrImport, " <module>[.<submodule>]*;')");
```

Comment:
Not a big fan of a grammar description like this in the message; in my opinion, it's not very user-friendly. An inexperienced user will likely not want to read a grammar[^1], whereas an experienced user will likely not need to be shown this error.

Was it there before?

[^1]: God knows I get scared by grammars on language feature description pages. I'm looking at you [Chapel](https://chapel-lang.org/docs/language/spec/tuples.html), [Coq](https://coq.inria.fr/refman/proof-engine/ltac.html).

Reply:
It was there before. I see your point about user friendliness. I can't think of a good replacement for it; I would be in favor of just removing this message with the example.